File: leetcode/easy/convert-1d-array-into-2d-array.py
# https://leetcode.com/problems/convert-1d-array-into-2d-array/

class Solution:
    def construct2DArray(self, ori: List[int], m: int, n: int) -> List[List[int]]:
        # MINE
        '''
        if len(ori) != m*n:
            return []

        # Python uses shallow lists
        # ans = [[0]*n]*m
        # the better way to declare a 2d array is: 
        ans = [[0 for i in range(n)] for j in range(m)]

        for i in range(len(ori)):
            ans[floor(i/n)][i%n] = ori[i]
            # print(floor(i/n), i%n, ori[i], ans)
        return ans
        '''

        # ANSWER
        ans = []
        if len(ori) == m*n:
            for i in range(0, len(ori), n):
                ans.append(ori[i:i+n])
        return ans

File: leetcode/easy/convert-1d-array-into-2d-array.ts
// https://leetcode.com/problems/convert-1d-array-into-2d-array/

// MINE
/*
function construct2DArray(ori: number[], m: number, n: number): number[][] {
    const ans = []
    let i = 0
    for(let row=0; row<m; ++row) {
        const arr = []
        for(let col=0; col<n; ++col) {
            if(!ori[i+col]) return []
            arr.push(ori[i+col])
        }
        ans.push(arr)
        i += n
    }
    if(i < ori.length) return []
    return ans
};
*/

// ANSWER: O(n)
function construct2DArray(ori: number[], m: number, n: number): number[][] {
  if (ori.length !== m * n) return []
  const ans = new Array(m).fill(0).map(_ => new Array(n))
  for (let i = 0; i < ori.length; ++i) ans[Math.floor(i / n)][i % n] = ori[i]
  return ans
}
